flowkit_jwt_generator/flowkit_jwt_generator/jwt.py

```diff
@@ -0,0 +1,229 @@
+
+    """
+    query_kinds = []
+    aggregation_types = set()
+    spec = requests.get(f"{flowapi_url}/api/0/spec/openapi.json").json()
+    for q_class, q_spec in spec["components"]["schemas"].items():
+        try:
+            query_kinds.append(q_spec["properties"]["query_kind"]["enum"][0])
+            aggregation_types.update(q_spec["properties"]["aggregation_unit"]["enum"])
+        except KeyError:
+            pass
+    query_kinds = sorted(query_kinds)
+    aggregation_types = sorted(aggregation_types)
+    all_claims = {
+        query_kind: {
+            "permissions": permissions_types,
+            "spatial_aggregation": aggregation_types,
+        }
+        for query_kind in query_kinds
+    }
+    all_claims.update(
+        {
+            "geography": {
+                "permissions": permissions_types,
+                "spatial_aggregation": aggregation_types,
+            },
+            "available_dates": {"permissions": {"get_result": True}},
+        }
+    )
+    return all_claims
+
+
+@pytest.fixture
+def universal_access_token(flowapi_url: str, access_token_builder: Callable) -> str:
+    """
+
+    Parameters
+    ----------
+    flowapi_url : str
+    access_token_builder : pytest.fixture
+
+    Returns
+    -------
+    str
+        The token
+
+    """
+    return access_token_builder(get_all_claims_from_flowapi(flowapi_url=flowapi_url))
+
+
+@click.group(chain=True)
+@click.argument("username", type=str)
+@click.argument("secret-key", type=str)
+@click.argument("lifetime", type=int)
+def print_token(username, secret_key, lifetime):
+    """
+    Generate a JWT token for access to FlowAPI.
+
+    Use the --all-access option to generate a token which allows full access to a
+    specific FlowAPI server, or specify the queries you want the token to grant
+    access to by providing multiple --query options.
+
+    For example:
+
+    generate-jwt TEST_USER SECRET 1 --all-access http://localhost:9090
+
+    Or,
+
+    generate-jwt TEST_USER SECRET 1 --query -a admin0 -a admin1 -p run -p get_result daily_location --query -a admin0 -p get_result flows
+    """
+    pass
+
+
+@print_token.resultcallback()
+def output_token(claims, username, secret_key, lifetime):
+    click.echo(
+        generate_token(
+            username=username,
+            secret=secret_key,
+            lifetime=datetime.timedelta(days=lifetime),
+            claims=dict(ChainMap(*claims)),
+        )
+    )
+
+
+@print_token.command("--query")
+@click.argument("query_name", type=str)
+@click.option(
+    "--permission",
+    "-p",
+    type=click.Choice(permissions_types.keys()),
+    multiple=True,
+    help="Types of access allowed.",
+)
+@click.option(
+    "--aggregation",
+    "-a",
+    type=click.Choice(aggregation_types),
+    multiple=True,
+    help="Spatial aggregation level of access allowed.",
+)
+def named_query(query_name, permission, aggregation):
+    if len(permission) == 0 and len(aggregation) == 0:
```

Comment:
Should this be an `or` rather than an `and`? Otherwise it's possible to generate tokens which have no run/poll/get_result permissions at all (only aggregation units), and vice versa.